m3/src/main/java/com/uber/m3/tally/m3/ResourcePool.java

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.uber.m3.tally.m3;
+
+import com.uber.m3.thrift.generated.CountValue;
+import com.uber.m3.thrift.generated.GaugeValue;
+import com.uber.m3.thrift.generated.Metric;
+import com.uber.m3.thrift.generated.MetricBatch;
+import com.uber.m3.thrift.generated.MetricTag;
+import com.uber.m3.thrift.generated.MetricValue;
+import com.uber.m3.thrift.generated.TimerValue;
+import com.uber.m3.util.Construct;
+import com.uber.m3.util.ObjectPool;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolFactory;
+
+import java.util.HashSet;
+import java.util.List;
+
+/**
+ * A collection of {@link ObjectPool}s used for the {@link M3Reporter}.
+ */
+public class ResourcePool {
```

Comment:
Could you write a benchmark and validate resource pooling is more performant than relying on the JVM to construct/remove short-lived objects? 

JVM runtime >> Go's runtime, so I don't think we can port the patterns/idioms directly

Reply:
Great idea. I hooked up [JMH benchmarking](http://openjdk.java.net/projects/code-tools/jmh/) to test this. I benchmarked:

1. Initialized a ResourcePool, took one object of each type from the pool and immediately released it back. I repeated this 10,000 times, suggested garbage collection via `System.gc()`, then repeated this four more times.

1. Naively instantiated each ResourcePool object type via `new` 10,000 times, again suggested garbage collection, then repeated four more times.

**Results:**
```
Benchmark                                        Mode  Cnt   Score   Error  Units
ResourcePoolBenchmark.usingResourcePool         thrpt    5  19.671 ± 0.617  ops/s
ResourcePoolBenchmark.usingNaiveObjectCreation  thrpt    5  40.551 ± 1.684  ops/s
```

These results suggest removing all the `Cached*` constructs and replace them with naively instantiating objects and letting GC do its magic. Any last comments before I start ripping things out? 😐

Reply:
Nice find. Only concern is if the benchmark emulates the workload - do we only have short lived objects being pulled?

Reply:
Yes. The pool is essentially for creating Metrics, and the Processor loop should very quickly (in a separate thread) pick them up from the queue and release them back to the pool.